### PR TITLE
Fix issues with taking screenshots during integration test failures

### PR DESCRIPTION
**Pull request by tmeschter**
1. Don't try to take screenshots when there is no screen.
2. If an exception occurs in our `FirstChanceException` handler, avoid recursively entering the handler and overflowing the stack.

### CODE REVIEW

File: src/VisualStudio/IntegrationTest/TestUtilities/ScreenshotService.cs [hunk 1]
```diff
@@ -28,12 +33,23 @@ public static void TakeScreenshot(string fullPath)
...
+        /// </returns>
+        private static Bitmap TryCaptureFullScreen()
         {
-            var bitmap = new Bitmap(
-                width: Screen.PrimaryScreen.Bounds.Width,
-                height: Screen.PrimaryScreen.Bounds.Height);
+            int width = Screen.PrimaryScreen.Bounds.Width;
+            int height = Screen.PrimaryScreen.Bounds.Height;
+
+            if (width < 0 || height < 0)
```
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/TestUtilities/ScreenshotService.cs]: Should this be <= in case if they come back zero?

Comment (tmeschter) [inline src/VisualStudio/IntegrationTest/TestUtilities/ScreenshotService.cs]: Blah. <= is what I meant; clearly my fingers didn't get the message while I was typing.

**Comment by tmeschter** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest please

**Comment by tmeschter** [comment]
@dotnet-bot test windows_debug_vs-integration_prtest  please

**Comment by tmeschter** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by tmeschter** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

File: src/VisualStudio/IntegrationTest/TestUtilities/ScreenshotService.cs [hunk 2]
```diff
@@ -28,12 +33,23 @@ public static void TakeScreenshot(string fullPath)
...
         {
-            var bitmap = new Bitmap(
-                width: Screen.PrimaryScreen.Bounds.Width,
-                height: Screen.PrimaryScreen.Bounds.Height);
+            int width = Screen.PrimaryScreen.Bounds.Width;
+            int height = Screen.PrimaryScreen.Bounds.Height;
+
+            if (width <= 0 || height <= 0)
+            {
+                // Don't try to take a screenshot if there is no screen.
```
Comment (tannergooding) [inline src/VisualStudio/IntegrationTest/TestUtilities/ScreenshotService.cs]: We might want to fail here. There are many other issues that could arise if we don't have an interactive session (Any tests for whether a window is visible comes to mind).

File: src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs [hunk 3]
```diff
@@ -21,6 +21,9 @@ namespace Microsoft.VisualStudio.IntegrationTest.Utilities
 {
     public sealed class VisualStudioInstanceFactory : IDisposable
     {
+        [ThreadStatic]
+        private static bool s_inHandler;
```
Comment (tannergooding) [inline src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs]: Isn't `ThreadLocal<T>` generally preferred (or is that only when you need to supply an initializer)?

File: src/VisualStudio/IntegrationTest/TestUtilities/ScreenshotService.cs (hunk 2 above)
Comment (tmeschter) [inline src/VisualStudio/IntegrationTest/TestUtilities/ScreenshotService.cs]: So something has already gone wrong, and as a side effect of that we're trying to take a screenshot. I don't really want to fail the test while taking the screenshot as that is a distraction from the real problem.

Comment (tannergooding) [inline src/VisualStudio/IntegrationTest/TestUtilities/ScreenshotService.cs]: We could always do what the legacy tests did and validate we have a session before starting a test.

Comment (tannergooding) [inline src/VisualStudio/IntegrationTest/TestUtilities/ScreenshotService.cs]: Not particular about it either way though.

File: src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs (hunk 3 above)
Comment (tmeschter) [inline src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs]: You _have_ to use `ThreadLocal<T>` if you need to supply an initializer. `ThreadLocal<T>` also needs to be disposed. That wouldn't be a problem here, but `ThreadStatic` is sufficient for our purposes.

**Pull request merged**